README
NAME
    WebService::Recruit::Dokoiku - A Interface for Dokoiku Web Service Beta

SYNOPSIS
        use WebService::Recruit::Dokoiku;

        my $doko = WebService::Recruit::Dokoiku->new();
        $doko->key( 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx' );

        my $param = {
            lat_jgd =>  '35.6686',
            lon_jgd =>  '139.7593',
            name    =>  'ATM',
        };
        my $res = $doko->searchPOI( %$param );
        my $list = $res->root->poi;
        foreach my $poi ( @$list ) {
            print "name: ", $poi->name, "\n";
            print "addr: ", $poi->address, "\n";
            print "web:  ", $poi->dokopcurl, "\n";
            print "map:  ", $poi->dokomapurl, "\n";
            print "\n";
        }

DESCRIPTION
    This module is a interface for the Dokoiku Web Service *Beta*, provided
    by Recruit Co., Ltd., Japan. It provides three API methods: "searchPOI",
    "getLandmark" and "getStation". With these methods, you can find almost
    all of shops, restaurants and many other places in Japan.

   searchPOI
    See WebService::Recruit::Dokoiku::SearchPOI for details.

        my $res = $doko->searchPOI( lmcode => 4212, name => 'ATM' );

   getLandmark
    See WebService::Recruit::Dokoiku::GetLandmark for details.

        my $res = $doko->getLandmark( name => 'SHIBUYA 109' );

   getStation
    See WebService::Recruit::Dokoiku::GetStation for details.

        my $res = $doko->getStation( lon_jgd => 139.758, lat_jgd => 35.666 );

  PAGING
    Each API response also provides paging methods following:

   page
    "page" method returns a Data::Page instance.

        my $page = $res->page();
        print "Total: ", $page->total_entries, "\n";
        print "Page: ", $page->current_page, "\n";
        print "Last: ", $page->last_page, "\n";

   pageset
    "pageset" method returns a Data::Pageset instance

        my $pageset = $res->pageset( 'fixed' );
        $pageset->pages_per_set($pages_per_set);
        my $set = $pageset->pages_in_set();
        foreach my $num ( @$set ) {
            print "$num ";
        }

   page_param
    "page_param" method returns a hash to specify the page for the next
    request.

        my %hash  = $res->page_param( $page->next_page );

   page_query
    "page_query" method returns a query string to specify the page for the
    next request.

        my $query = $res->page_query( $page->prev_page );

  TreePP CONFIG
    This modules uses XML::TreePP module internally. Following methods are
    available to configure it.

        my $doko = WebService::Recruit::Dokoiku->new();
        $doko->utf8_flag( 1 );
        $doko->user_agent( 'Foo-Bar/1.0 ' );
        $doko->lwp_useragent( LWP::UserAgent->new() );
        $doko->http_lite( HTTP::Lite->new() );

SEE ALSO
    http://www.doko.jp/api/

AUTHOR
    Yusuke Kawasaki <http://www.kawa.net/>

    This module is unofficial and released by the authour in person.

COPYRIGHT AND LICENSE
    Copyright (c) 2007 Yusuke Kawasaki. All rights reserved. This program is
    free software; you can redistribute it and/or modify it under the same
    terms as Perl itself.



Changes
# WebService::Recruit::Dokoiku Changes

2007/08/16 version 0.10:
    - Class::Accessor::Children::Fast required
    - WebService::Recruit::Dokoiku::Base splitted
    - WebService::Recruit::Dokoiku::GetStation typo fixed
    - response accessors read only

2007/08/13 version 0.07:
    - the first release!

# http://www.kawa.net/works/recruit/webservice/dokoiku.html


lib/WebService/Recruit/Dokoiku/SearchPOI.pm
package WebService::Recruit::Dokoiku::SearchPOI;
use strict;
use vars qw( $VERSION );
use base qw( WebService::Recruit::Dokoiku::API );
$VERSION = '0.07';

sub url { 'http://api.doko.jp/v1/searchPOI.do'; }
sub force_array { [qw( poi )]; }
sub elem_class { 'WebService::Recruit::Dokoiku::SearchPOI::Element'; }
sub query_class { 'WebService::Recruit::Dokoiku::SearchPOI::Query'; }

package WebService::Recruit::Dokoiku::SearchPOI::Query;
use strict;
use base qw( Class::Accessor::Fast );
__PACKAGE__->mk_accessors(qw(
    key format callback pagenum pagesize keyword name tel
    lat_jgd lon_jgd radius lmcode iarea order
));

package WebService::Recruit::Dokoiku::SearchPOI::Element;
use strict;
use base qw( Class::Accessor::Fast );
__PACKAGE__->mk_accessors(qw( results ));

package WebService::Recruit::Dokoiku::SearchPOI::Element::results;
use strict;
use base qw( Class::Accessor::Fast );
__PACKAGE__->mk_accessors(qw( status totalcount pagenum poi ));

package WebService::Recruit::Dokoiku::SearchPOI::Element::poi;
use strict;
use base qw( Class::Accessor::Fast );
__PACKAGE__->mk_accessors(qw( code name kana tel address stationcode
    station distance dokopcurl dokomburl dokomapurl reviewrank tag ));

=head1 NAME

WebService::Recruit::Dokoiku::SearchPOI - Dokoiku Web Service "searchPOI" API

=head1 SYNOPSIS

    use WebService::Recruit::Dokoiku;

    my $doko = WebService::Recruit::Dokoiku->new();
    $doko->key( 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx' );

    my $param = {
        lat_jgd =>  '35.6686',
        lon_jgd =>  '139.7593',
        name    =>  'ATM',
    };
    my $res = $doko->searchPOI( %$param );
    die "error!" if $res->is_error;

    my $list = $res->root->poi;
    foreach my $poi ( @$list ) {
        print "name: ", $poi->name, "\n";
        print "addr: ", $poi->address, "\n";
        print "web:  ", $poi->dokopcurl, "\n";
        print "map:  ", $poi->dokomapurl, "\n";
        print "\n";
    }

    my $root = $res->root;

=head1 DESCRIPTION

The request to this API requires one or some of query parameters below:

    my $param = {
        pagenum     =>  '1',
        pagesize    =>  '10',
        keyword     =>  'keyword for place',
        name        =>  'name of place',
        tel         =>  '03-3575-1111',
        lat_jgd     =>  '35.6686',
        lon_jgd     =>  '139.7593',
        radius      =>  '1000',
        lmcode      =>  '1908',
        iarea       =>  '05800',
        order       =>  '1',
    };

The response from this API is tree structured and provides methods below:

    $root->status
    $root->totalcount
    $root->pagenum
    $root->poi->[0]->code
    $root->poi->[0]->name
    $root->poi->[0]->kana
    $root->poi->[0]->tel
    $root->poi->[0]->address
    $root->poi->[0]->stationcode
    $root->poi->[0]->station
    $root->poi->[0]->distance
    $root->poi->[0]->dokopcurl
    $root->poi->[0]->dokomburl
    $root->poi->[0]->dokomapurl
    $root->poi->[0]->reviewrank
    $root->poi->[0]->tag

And paging methods are provided, see L<WebService::Recruit::Dokoiku/PAGING>.
This module is based on L<XML::OverHTTP>.

=head1 SEE ALSO

L<WebService::Recruit::Dokoiku>

=head1 AUTHOR

Yusuke Kawasaki L<http://www.kawa.net/>

This module is unofficial and released by the authour in person.

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2007 Yusuke Kawasaki. All rights reserved.
This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=cut
1;


lib/WebService/Recruit/Dokoiku/GetStation.pm
package WebService::Recruit::Dokoiku::GetStation;
use strict;
use vars qw( $VERSION );
use base qw( WebService::Recruit::Dokoiku::API );
$VERSION = '0.07';

sub url { 'http://api.doko.jp/v1/getStation.do'; }
sub force_array { [qw( landmark )]; }
sub elem_class { 'WebService::Recruit::Dokoiku::GetStation::Element'; }
sub query_class { 'WebService::Recruit::Dokoiku::GetStation::Query'; }

package WebService::Recruit::Dokoiku::GetStation::Query;
use strict;
use base qw( Class::Accessor::Fast );
__PACKAGE__->mk_accessors(qw(
    key format callback pagenum pagesize name code lat_jgd lat_jgd radius iarea
));

package WebService::Recruit::Dokoiku::GetStation::Element;
use strict;
use base qw( Class::Accessor::Fast );
__PACKAGE__->mk_accessors(qw( results ));

package WebService::Recruit::Dokoiku::GetStation::Element::results;
use strict;
use base qw( Class::Accessor::Fast );
__PACKAGE__->mk_accessors(qw( status totalcount pagenum landmark ));

package WebService::Recruit::Dokoiku::GetStation::Element::landmark;
use strict;
use base qw( Class::Accessor::Fast );
__PACKAGE__->mk_accessors(qw(
    code name dokopcurl dokomburl dokomapurl lat_jgd lon_jgd lat_tky lon_tky
));

=head1 NAME

WebService::Recruit::Dokoiku::GetStation - Dokoiku Web Service "getStation" API

=head1 SYNOPSIS

    use WebService::Recruit::Dokoiku;

    my $doko = WebService::Recruit::Dokoiku->new();
    $doko->key( 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx' );

    my $param = {
        lon_jgd     =>  139.758,
        lat_jgd     =>   35.666,
    };
    my $res = $doko->getStation( %$param );
    die "error!" if $res->is_error;

    my $list = $res->root->landmark;
    foreach my $landmark ( @$list ) {
        print "code: ", $landmark->code, "\n";
        print "name: ", $landmark->name, "\n";
        print "web:  ", $landmark->dokopcurl, "\n";
        print "map:  ", $landmark->dokomapurl, "\n";
        print "\n";
    }

    my $root = $res->root;

=head1 DESCRIPTION

The request to this API requires one or some of query parameters below:

    my $param = {
        pagenum     =>  '1',
        pagesize    =>  '10',
        name        =>  'name of station',
        code        =>  '4254',
        lat_jgd     =>  '35.6686',
        lon_jgd     =>  '139.7593',
        radius      =>  '1000',
        iarea       =>  '05800',
    };

The response from this API is tree structured and provides methods below:

    $root->status
    $root->totalcount
    $root->pagenum
    $root->landmark->[0]->code
    $root->landmark->[0]->name
    $root->landmark->[0]->dokopcurl
    $root->landmark->[0]->dokomburl
    $root->landmark->[0]->dokomapurl
    $root->landmark->[0]->lat_jgd
    $root->landmark->[0]->lon_jgd
    $root->landmark->[0]->lat_tky
    $root->landmark->[0]->lon_tky

And paging methods are provided, see L<WebService::Recruit::Dokoiku/PAGING>.
This module is based on L<XML::OverHTTP>.

=head1 SEE ALSO

L<WebService::Recruit::Dokoiku>

=head1 AUTHOR

Yusuke Kawasaki L<http://www.kawa.net/>

This module is unofficial and released by the authour in person.

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2007 Yusuke Kawasaki. All rights reserved.
This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=cut
1;


lib/WebService/Recruit/Dokoiku/GetLandmark.pm
package WebService::Recruit::Dokoiku::GetLandmark;
use strict;
use vars qw( $VERSION );
use base qw( WebService::Recruit::Dokoiku::API );
$VERSION = '0.07';

sub url { 'http://api.doko.jp/v1/getLandmark.do'; }
sub force_array { [qw( landmark )]; }
sub elem_class { 'WebService::Recruit::Dokoiku::GetLandmark::Element'; }
sub query_class { 'WebService::Recruit::Dokoiku::GetLandmark::Query'; }

package WebService::Recruit::Dokoiku::GetLandmark::Query;
use strict;
use base qw( Class::Accessor::Fast );
__PACKAGE__->mk_accessors(qw(
    key format callback pagenum pagesize name code lat_jgd lat_jgd radius iarea
));

package WebService::Recruit::Dokoiku::GetLandmark::Element;
use strict;
use base qw( Class::Accessor::Fast );
__PACKAGE__->mk_accessors(qw( results ));

package WebService::Recruit::Dokoiku::GetLandmark::Element::results;
use strict;
use base qw( Class::Accessor::Fast );
__PACKAGE__->mk_accessors(qw( status totalcount pagenum landmark ));

package WebService::Recruit::Dokoiku::GetLandmark::Element::landmark;
use strict;
use base qw( Class::Accessor::Fast );
__PACKAGE__->mk_accessors(qw( code name dokopcurl dokomburl dokomapurl ));

=head1 NAME

WebService::Recruit::Dokoiku::GetLandmark - Dokoiku Web Service "getLandmark" API

=head1 SYNOPSIS

    use WebService::Recruit::Dokoiku;

    my $doko = WebService::Recruit::Dokoiku->new();
    $doko->key( 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx' );

    my $param = {
        name    =>  'SHIBUYA109',
    };
    my $res = $doko->getLandmark( %$param );
    die "error!" if $res->is_error;

    my $list = $res->root->landmark;
    foreach my $landmark ( @$list ) {
        print "code: ", $landmark->code, "\n";
        print "name: ", $landmark->name, "\n";
        print "web:  ", $landmark->dokopcurl, "\n";
        print "map:  ", $landmark->dokomapurl, "\n";
        print "\n";
    }

    my $root = $res->root;

=head1 DESCRIPTION

The request to this API requires one or some of query parameters below:

    my $param = {
        pagenum     =>  '1',
        pagesize    =>  '10',
        name        =>  'name of station',
        code        =>  '4254',
        lat_jgd     =>  '35.6686',
        lon_jgd     =>  '139.7593',
        radius      =>  '1000',
        iarea       =>  '05800',
    };

The response from this API is tree structured and provides methods below:

    $root->status
    $root->totalcount
    $root->pagenum
    $root->landmark->[0]->code
    $root->landmark->[0]->name
    $root->landmark->[0]->dokopcurl
    $root->landmark->[0]->dokomburl
    $root->landmark->[0]->dokomapurl

And paging methods are provided, see L<WebService::Recruit::Dokoiku/PAGING>.
This module is based on L<XML::OverHTTP>.

=head1 SEE ALSO

L<WebService::Recruit::Dokoiku>

=head1 AUTHOR

Yusuke Kawasaki L<http://www.kawa.net/>

This module is unofficial and released by the authour in person.

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2007 Yusuke Kawasaki. All rights reserved.
This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=cut
1;


lib/WebService/Recruit/Dokoiku.pm
package WebService::Recruit::Dokoiku;
use strict;
use base qw( Class::Accessor::Fast );
use vars qw( $VERSION );
$VERSION = '0.10';

use WebService::Recruit::Dokoiku::SearchPOI;
use WebService::Recruit::Dokoiku::GetLandmark;
use WebService::Recruit::Dokoiku::GetStation;

my $PARAMS = [qw( key pagesize )];
my $TPPCFG = [qw( user_agent lwp_useragent http_lite utf8_flag )];
__PACKAGE__->mk_accessors( @$PARAMS, @$TPPCFG );

sub new {
    my $package = shift;
    my $self    = {@_};
    $self->{user_agent} ||= __PACKAGE__."/$VERSION ";
    bless $self, $package;
    $self;
}

sub init_treepp_config {
    my $self = shift;
    my $api  = shift;
    my $treepp = $api->treepp();
    foreach my $key ( @$TPPCFG ) {
        next unless exists $self->{$key};
        next unless defined $self->{$key};
        $treepp->set( $key => $self->{$key} );
    }
}

sub init_query_param {
    my $self = shift;
    my $api  = shift;
    foreach my $key ( @$PARAMS ) {
        next unless exists $self->{$key};
        next unless defined $self->{$key};
        $api->add_param( $key => $self->{$key} );
    }
}

sub searchPOI {
    my $self = shift or return;
    $self = $self->new() unless ref $self;
    my $api = WebService::Recruit::Dokoiku::SearchPOI->new();
    $self->init_treepp_config( $api );
    $self->init_query_param( $api );
    $api->add_param( @_ );
    $api->request();
    $api;
}

sub getLandmark {
    my $self = shift or return;
    $self = $self->new() unless ref $self;
    my $api = WebService::Recruit::Dokoiku::GetLandmark->new();
    $self->init_treepp_config( $api );
    $self->init_query_param( $api );
    $api->add_param( @_ );
    $api->request();
    $api;
}

sub getStation {
    my $self = shift or return;
    $self = $self->new() unless ref $self;
    my $api = WebService::Recruit::Dokoiku::GetStation->new();
    $self->init_treepp_config( $api );
    $self->init_query_param( $api );
    $api->add_param( @_ );
    $api->request();
    $api;
}

=head1 NAME

WebService::Recruit::Dokoiku - A Interface for Dokoiku Web Service Beta

=head1 SYNOPSIS

    use WebService::Recruit::Dokoiku;

    my $doko = WebService::Recruit::Dokoiku->new();
    $doko->key( 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx' );

    my $param = {
        lat_jgd =>  '35.6686',
        lon_jgd =>  '139.7593',
        name    =>  'ATM',
    };
    my $res = $doko->searchPOI( %$param );
    my $list = $res->root->poi;
    foreach my $poi ( @$list ) {
        print "name: ", $poi->name, "\n";
        print "addr: ", $poi->address, "\n";
        print "web:  ", $poi->dokopcurl, "\n";
        print "map:  ", $poi->dokomapurl, "\n";
        print "\n";
    }

=head1 DESCRIPTION

This module is a interface for the Dokoiku Web Service I<Beta>,
provided by Recruit Co., Ltd., Japan.
It provides three API methods: L</searchPOI>, L</getLandmark>
and L</getStation>.
With these methods, you can find almost all of shops, restaurants
and many other places in Japan.

=head1 METHODS

=head2 new

This is the constructor method for this class.

    my $doko = WebService::Recruit::Dokoiku->new();

This accepts optional parameters.

    my $conf = { key => 'your_auth_key', utf8_flag => 1 };
    my $doko = WebService::Recruit::Dokoiku->new( %$conf );

=head2 key

A valid developer key is required to make a request.

    $doko->key( 'your_auth_key' );

=head2 searchPOI

This makes a request for C<searchPOI> API.
See L<WebService::Recruit::Dokoiku::SearchPOI> for details.

    my $res = $doko->searchPOI( lmcode => 4212, name => 'ATM' );

=head2 getLandmark

This makes a request for C<getLandmark> API.
See L<WebService::Recruit::Dokoiku::GetLandmark> for details.

    my $res = $doko->getLandmark( name => 'SHIBUYA 109' );

=head2 getStation

This makes a request for C<getStation> API.
See L<WebService::Recruit::Dokoiku::GetStation> for details.

    my $res = $doko->getStation( lon_jgd => 139.758, lat_jgd => 35.666 );

=head2 utf8_flag / user_agent / lwp_useragent / http_lite

This modules uses L<XML::TreePP> module internally.
Following methods are available to configure it.

    $doko->utf8_flag( 1 );
    $doko->user_agent( 'Foo-Bar/1.0 ' );
    $doko->lwp_useragent( LWP::UserAgent->new() );
    $doko->http_lite( HTTP::Lite->new() );

=head1 SEE ALSO

http://www.doko.jp/api/

=head1 AUTHOR

Yusuke Kawasaki L<http://www.kawa.net/>

This module is unofficial and released by the authour in person.

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2007 Yusuke Kawasaki. All rights reserved.
This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=cut
1;
